### PR TITLE
Edit announcement and contact card

### PR DESCRIPTION
**Pull request by theo-auffeuvre**
Making announcement card and contact edition working with error validation 

Some other small fixes

### CODE REVIEW

**Comment by Samuel-Therrien-Beslogic** [comment]
The amount of changes in api.ts doesn't look right. Can you regenerate it using the steps in the README ?

It's also odd that you got that many package-lock.json changes without any package.json change

I see requirements changes in the backend requirements that don't make sense there. I don't see any change to the pyproject or dev requirements.

The UV branch was also merged whilst failing to even install on the CI.

**Comment by Samuel-Therrien-Beslogic** [comment]
Linting error looks like it's caused by an updated rule. So probably caused by the package-lock.json  changes
­> Configuration for rule "@typescript-eslint/prefer-string-starts-ends-with" is invalid:
> 	Value [{"allowSingleElementEquality":"always"}] should NOT have more than 0 items.

File: canopeum_backend/canopeum_backend/views.py [hunk 1]
```diff
@@ -731,11 +731,12 @@ def patch(self, request: Request, siteId):
         try:
             announcement = Announcement.objects.get(site=siteId)
         except Announcement.DoesNotExist:
-            return Response(status=status.HTTP_404_NOT_FOUND)
+            announcement = Announcement.objects.create()
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/views.py:734]:
```py
        announcement = Announcement.objects.get_or_create(site=siteId)
```

File: canopeum_backend/canopeum_backend/settings.py [hunk 2]
```diff
@@ -171,20 +171,15 @@ def get_secret(key: str, default: str):
...
-# }
 DATABASES = {
     "default": dj_database_url.parse(
         "mysql://canopeum_user:{}@{}:3308/canopeum_db".format(
             get_secret("MYSQL_PASSWORD_CANOPEUM", ""),
             os.getenv("MYSQL_HOST_CANOPEUM", "localhost"),
         ),
+        # "default": dj_database_url.parse(
+        #     f"mysql://canopeum_user:{get_secret("MYSQL_PASSWORD_CANOPEUM", "")}
+        # @{get_secret("MYSQL_HOST_CANOPEUM", "localhost")}:3306/canopeum_db",
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/settings.py:182]: Are these changes intentional? Even uncommented this doesn't make much sense

Comment (theo-auffeuvre) [reply canopeum_backend/canopeum_backend/settings.py]: MAde it clearier

File: canopeum_backend/canopeum_backend/views.py (hunk 1 above)
Comment (theo-auffeuvre) [reply canopeum_backend/canopeum_backend/views.py]: Done!

File: canopeum_frontend/src/components/analytics/BatchTable.tsx [hunk 3]
```diff
@@ -106,8 +106,8 @@ const BatchTable = (props: Props) => {
                 style={{ borderColor: cellBorderColor }}
               >
                 <ul>
-                  {batch.species.map(type => (
-                    <li key={`batch-${batch.id}-treeType-${type.en}`}>
+                  {batch.species.map((type, i) => (
+                    <li key={`batch-${batch.id}-treeType-${type.en}-${i}`}>
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/components/analytics/BatchTable.tsx:110]: Is there some other unique identifier to the species we can use here? The point is to help React know what and when to re-render, if you swap the order of two items that had the same ID but differentiated on their index, then they'll end up with the same key and not swap visually in the UI because React will think it doesn't need to re-render !
https://react.dev/learn/rendering-lists#where-to-get-your-key
https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-array-index-key.md

File: canopeum_frontend/src/main.tsx [hunk 4]
```diff
@@ -26,8 +25,6 @@ if (root == null) {
   throw new Error('Could not find element #root to bootstrap React.')
 } else {
   createRoot(root).render(
-    <StrictMode>
-      <App />
-    </StrictMode>,
+    <App />,
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/main.tsx:28]: Why remove the StrictMode ? This change hides runtime issues during development

File: canopeum_frontend/src/pages/SiteSocialPage.tsx [hunk 5]
```diff
@@ -139,8 +139,18 @@ const SiteSocialPage = () => {
...
+                announcement={site.announcement}
+                viewMode={viewMode}
+                onEdit={(editedAnnoucement: Announcement) =>
+                  setSite(site => ({ ...site, announcement: editedAnnoucement } as SiteSocial))}
+              />
+              <ContactCard
+                contact={site.contact}
+                viewMode={viewMode}
+                onEdit={(editedContact: Contact) =>
+                  setSite(site => ({ ...site, contact: editedContact } as SiteSocial))}
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/pages/SiteSocialPage.tsx:152]: Unless it shadows a variable, you can shorten to
```suggestion
                onEdit={(announcement: Announcement) =>
                  setSite(site => ({ ...site, announcement} as SiteSocial))}
              />
              <ContactCard
                contact={site.contact}
                viewMode={viewMode}
                onEdit={(contact: Contact) =>
                  setSite(site => ({ ...site, contact} as SiteSocial))}
```

Type should be infered from the assigned parameter, it's likely you can do:

```suggestion
                onEdit={announcement => setSite(site => ({ ...site, announcement} as SiteSocial))}
              />
              <ContactCard
                contact={site.contact}
                viewMode={viewMode}
                onEdit={contact => setSite(site => ({ ...site, contact} as SiteSocial))}
```

I'm not sure about that cast either.

File: docker-compose.yml [hunk 6]
```diff
@@ -27,6 +27,7 @@ services:
     environment:
       MYSQL_DATABASE: canopeum_db
       MYSQL_HOST_CANOPEUM: /run/secrets/MYSQL_HOST_CANOPEUM
+      MYSQL_PORT_CANOPEUM: /run/secrets/MYSQL_PORT_CANOPEUM
```
Comment (Samuel-Therrien-Beslogic) [inline docker-compose.yml:30]: Setup changes are preferable as a different PR. If simple enough I can almost immediately approve those. (and I agree with this change, more control for the local dev and it doesn't hurt anything)

File: canopeum_backend/canopeum_backend/views.py [hunk 7]
```diff
@@ -728,14 +728,12 @@ class AnnouncementDetailAPIView(APIView):
...
-            announcement = Announcement.objects.get(site=siteId)
-        except Announcement.DoesNotExist:
-            return Response(status=status.HTTP_404_NOT_FOUND)
+        announcement = Announcement.objects.get_or_create(site=siteId)
 
-        serializer = AnnouncementSerializer(announcement, data=request.data)
+        serializer = AnnouncementSerializer(announcement[0], data=request.data)
         if serializer.is_valid():
             serializer.save()
+            Site.objects.filter(pk=siteId).update(announcement=announcement[0])
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/views.py:736]: Can't comment across deleted lines, but here I think it'd make more sense to index once (this goes with the above comment)

```suggestion
        serializer = AnnouncementSerializer(announcement, data=request.data)
        if serializer.is_valid():
            serializer.save()
            Site.objects.filter(pk=siteId).update(announcement=announcement)
```

File: canopeum_backend/canopeum_backend/views.py [hunk 8]
```diff
@@ -728,14 +728,12 @@ class AnnouncementDetailAPIView(APIView):
         operation_id="announcement_update",
     )
     def patch(self, request: Request, siteId):
-        try:
-            announcement = Announcement.objects.get(site=siteId)
-        except Announcement.DoesNotExist:
-            return Response(status=status.HTTP_404_NOT_FOUND)
+        announcement = Announcement.objects.get_or_create(site=siteId)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/views.py:731]: Can't comment across deleted lines, but here I think it'd make more sense to index once (this goes with the below comment)
```suggestion
        announcement = Announcement.objects.get_or_create(site=siteId)[0]
```

File: canopeum_frontend/src/components/analytics/site-modal/SiteModal.tsx [hunk 9]
```diff
@@ -80,6 +81,7 @@ const SiteModal = ({ open, handleClose, siteId }: Props) => {
       researchPartner: siteDetail.researchPartnership,
       visibleOnMap: siteDetail.visibleMap,
     })
+    console.log(site.species)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/components/analytics/site-modal/SiteModal.tsx:84]:
```suggestion
```

File: canopeum_frontend/src/components/analytics/site-modal/SiteModal.tsx [hunk 10]
```diff
@@ -59,6 +59,7 @@ const SiteModal = ({ open, handleClose, siteId }: Props) => {
     }
 
     const siteDetail = await getApiClient().siteClient.detail(siteId)
+    console.log(siteDetail)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/components/analytics/site-modal/SiteModal.tsx:62]:
```suggestion
```

File: docker-compose.yml (hunk 6 above)
Comment (theo-auffeuvre) [reply docker-compose.yml]: See this [PR](https://github.com/BesLogic/releaf-canopeum/pull/204)

File: canopeum_frontend/src/components/analytics/BatchTable.tsx (hunk 3 above)
Comment (theo-auffeuvre) [reply canopeum_frontend/src/components/analytics/BatchTable.tsx]: You are true , it is not event necessary because a batch shouldn't have same treetype twice (was the case because of development in progress of edit batch)

File: canopeum_frontend/src/components/analytics/site-modal/SiteModal.tsx [hunk 11]
```diff
@@ -291,7 +292,7 @@ const SiteModal = ({ open, handleClose, siteId }: Props) => {
       <DialogActions>
         <button
           className='btn btn-outline-primary'
-          onClick={() => handleClose()}
+          onClick={() => handleClose('cancel')}
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/components/analytics/site-modal/SiteModal.tsx:295]: `undefined` was kindof implied as `'cancel'`. But explicit is better than implicit right? Maybe we should make the `reason` param required ?

File: canopeum_frontend/src/main.tsx (hunk 4 above)
Comment (theo-auffeuvre) [reply canopeum_frontend/src/main.tsx:28]: Don't really remember the reason I deleted it, I put it back

File: canopeum_frontend/src/components/analytics/site-modal/SiteModal.tsx (hunk 11 above)
Comment (theo-auffeuvre) [reply canopeum_frontend/src/components/analytics/site-modal/SiteModal.tsx:295]: image
I resolved it :(

File: canopeum_frontend/src/components/inputs/EmailTextField.tsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+      setError(null)
+      props.isValid(true)
+    }
+
+    props.onChange(value)
+  }
+
+  return (
+    <div className='d-flex flex-column flex-grow-1'>
+      {/* eslint-disable react/jsx-props-no-spreading -- Needed for custom input */}
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/components/inputs/EmailTextField.tsx:31]: We should consider re-configuring that one https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-props-no-spreading.md#rule-options

Probably `"html": "ignore"`. (no need to change anything in your PR, just taking a note)

File: canopeum_frontend/src/components/social/AnnouncementCard.tsx [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+  readonly announcement: Announcement,
+  readonly viewMode: PageViewMode,
+  readonly onEdit: (announcement: Announcement) => void,
+}
+
+const AnnouncementCard = ({ announcement, viewMode, onEdit }: Props) => {
+  const [isModalOpen, setIsModalOpen] = useState(false)
+
+  return (
+    <>
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/components/social/AnnouncementCard.tsx:18]: Small detail to reduce indentation
```suggestion
  return <>
```

File: canopeum_frontend/src/components/social/AnnouncementCard.tsx [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+      <SiteAnnouncementModal
+        announcement={announcement}
+        handleClose={(newAnnouncement: Announcement | null): void => {
+          setIsModalOpen(!isModalOpen)
+          if (newAnnouncement) onEdit(newAnnouncement)
+        }}
+        isOpen={isModalOpen}
+      />
+    </>
+  )
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/components/social/AnnouncementCard.tsx:54]: With comment above
```suggestion
  </>
```

File: canopeum_frontend/src/components/inputs/EmailTextField.tsx (hunk 12 above)
Comment (theo-auffeuvre) [reply canopeum_frontend/src/components/inputs/EmailTextField.tsx:31]: Ok so I keep theses comments.
For me, I'm balancing between keep the rule as is and ignore it. It give a good warn to the developer to be careful and give him the resposability to ignore it.

File: canopeum_frontend/src/components/social/site-modal/SiteAnnouncementModal.tsx [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+  const { openAlertSnackbar } = useContext(SnackbarContext)
+
+  const handleSubmitSiteAnnouncement = async () => {
+    try {
+      await getApiClient().announcementClient.update(
+        announcement.id,
+        editedAnnouncement as PatchedAnnouncement,
+      )
+    } catch (error: unknown) {
+      console.error(error)
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_frontend/src/components/social/site-modal/SiteAnnouncementModal.tsx:37]: This is a warning because the linter is telling you: "Maybe this error should be shown to the user instead of hidden in logs?". But we already use the snackbar. Maybe using a "console.error" for snackbar errors in general could be a good idea.

**Review by Samuel-Therrien-Beslogic — APPROVED** [review]

File: canopeum_frontend/src/components/inputs/EmailTextField.tsx (hunk 12 above)
Comment (Samuel-Therrien-Beslogic) [reply canopeum_frontend/src/components/inputs/EmailTextField.tsx:31]: Definitely a good warning, but I think there's cases where it can be beneficial for higher order components where we can reduce false-positives.
